src/Storages/StorageKeeperMap.cpp

```diff
@@ -0,0 +1,329 @@
+        auto res = zookeeper->tryGet(path, value);
+        if (!res)
+            return {};
+
+        MutableColumns columns(sample_block.cloneEmptyColumns());
+        insertRowForKey(columns, filter->filter, value);
+        processed = true;
+
+        return Chunk{std::move(columns), 1};
+    }
+
+    bool insertRowForKey(MutableColumns & columns, const std::string & key, const std::string & value)
+    {
+
+        for (size_t column_index = 0; column_index < column_names.size(); ++column_index)
+        {
+            if (column_names[column_index] == "key")
+                assert_cast<ColumnString &>(*columns[column_index]).insertData(key.data(), key.size());
+            else if (column_names[column_index] == "value")
+                assert_cast<ColumnString &>(*columns[column_index]).insertData(value.data(), value.size());
+        }
+        return true;
+    }
+
+public:
+    StorageKeeperMapSource(const Block & sample_block_, StorageKeeperMap & storage_, size_t max_block_size_, std::optional<KeyFilter> filter_)
+        : ISource(sample_block_)
+        , storage(storage_)
+        , sample_block(sample_block_.cloneEmpty())
+        , column_names(sample_block_.getNames())
+        , max_block_size(max_block_size_)
+        , filter(std::move(filter_))
+    {
+        has_value_column = std::any_of(column_names.begin(), column_names.end(), [](const auto & name) { return name == "value"; });
+
+        // TODO(antonio2368): Do it lazily in generate
+        if (!filter || filter->type != FilterType::EXACT)
+        {
+            auto zookeeper = storage.getClient();
+            keys = zookeeper->getChildren(storage.rootKeeperPath());
+        }
+    }
+
+    std::string getName() const override { return "StorageKeeperMapSource"; }
+
+    Chunk generate() override
+    {
+        if (filter && filter->type == FilterType::EXACT)
+            return generateSingleKey();
+
+        auto zookeeper = storage.getClient();
+
+        MutableColumns columns(sample_block.cloneEmptyColumns());
+        size_t num_rows = 0;
+        for (; num_rows < max_block_size && current_idx != keys.size(); ++current_idx)
+        {
+            const auto & key = keys[current_idx];
+            std::string value;
+            if (has_value_column)
+            {
+                auto path = fmt::format("{}/{}", storage.rootKeeperPath(), key);
+                auto res = zookeeper->tryGet(path, value);
+                if (!res)
+                    continue;
+            }
+
+            if (insertRowForKey(columns, key, value))
+                ++num_rows;
+        }
+
+        if (num_rows == 0)
+            return {};
+
+        return {std::move(columns), num_rows};
+    }
+};
+
+std::optional<KeyFilter> tryGetKeyFilter(const IAST & elem, const ContextPtr context)
+{
+    const auto * function = elem.as<ASTFunction>();
+    if (!function)
+        return std::nullopt;
+
+    if (function->name != "equals")
```

Comment:
It's possible to generalize it to support IN and some other expressions, maybe with KeyCondition.